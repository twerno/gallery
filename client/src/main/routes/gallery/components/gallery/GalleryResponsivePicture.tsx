import * as React from 'react';
import GalleryStyles from './styles/GalleryStyles';

export interface IResponsivePictureProps {
    alt: string;
    defaultImgSrc: string | undefined;
    responsiveSrcList?: Array<{ src: string | undefined; media: string; }>;
}

export interface IGalleryResponsivePictureProps {
    isVisible: boolean;
    onLoadHandler: () => void;
    pictureData: IResponsivePictureProps;
}

export const GalleryResponsivePicture = (props: IGalleryResponsivePictureProps) => {
    const { onLoadHandler, isVisible, pictureData } = props;

    const srcSet = pictureData.responsiveSrcList?.map(
        (src) => <source srcSet={src.src} media={src.media} key={`${src.src}_${src.media}`} />
    );

    return (
        <GalleryStyles.GalleryItemPicture
            onLoad={onLoadHandler}
            isHidden={!isVisible}
        >
            {srcSet}
            <GalleryStyles.GalleryItemImg src={pictureData.defaultImgSrc} alt={pictureData.alt} />
        </GalleryStyles.GalleryItemPicture>
    );
}