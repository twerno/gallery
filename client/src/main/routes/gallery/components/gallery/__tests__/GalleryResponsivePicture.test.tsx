import '@testing-library/jest-dom/extend-expect';
import { wait } from '@testing-library/react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import * as React from 'react';
import { componentBasicTests } from 'test/componentBasicTests';
import { GalleryResponsivePicture, IGalleryResponsivePictureProps } from '../GalleryResponsivePicture';

describe('GalleryResponsivePicture', () => {

    const props: IGalleryResponsivePictureProps = {
        pictureData: {
            alt: 'alt',
            defaultImgSrc: 'img.jpg',
            responsiveSrcList: [{ src: 'img-big.jpg', media: '(min-width: 600px)' }]
        },
        isVisible: true,
        onLoadHandler: () => { },
    };

    componentBasicTests(
        "GalleryResponsivePicture visible",
        <GalleryResponsivePicture {...{ ...props, isVisible: true }} />
    );

    componentBasicTests(
        "GalleryResponsivePicture hidden",
        <GalleryResponsivePicture {...{ ...props, isVisible: false }} />
    );


    test('onLoadHandler', async () => {

        const localProps: IGalleryResponsivePictureProps = {
            ...props,
            isVisible: false,
            onLoadHandler: jest.fn(),
        };

        const wrapper = shallow(
            <GalleryResponsivePicture {...localProps} />
        );
        const picture = wrapper.first();
        expect(picture.name()).toBe('styled.picture');

        await wait(() => wrapper.simulate('load'));

        expect(localProps.onLoadHandler).toBeCalled();
    });

});

