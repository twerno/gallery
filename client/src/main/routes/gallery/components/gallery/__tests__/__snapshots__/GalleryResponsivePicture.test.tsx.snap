// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryResponsivePicture [GalleryResponsivePicture hidden] Component basic tests matches the snapshot 1`] = `
.c0 {
  opacity: 1;
  position: relative;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
  min-width: 100%;
  position: absolute;
  opacity: 0;
}

.c1 {
  object-fit: cover;
  min-width: 100%;
  height: 100%;
}

<picture
  className="c0"
  onLoad={[Function]}
>
  <source
    media="(min-width: 600px)"
    srcSet="img-big.jpg"
  />
  <img
    alt="alt"
    className="c1"
    src="img.jpg"
  />
</picture>
`;

exports[`GalleryResponsivePicture [GalleryResponsivePicture visible] Component basic tests matches the snapshot 1`] = `
.c0 {
  opacity: 1;
  position: relative;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
  min-width: 100%;
}

.c1 {
  object-fit: cover;
  min-width: 100%;
  height: 100%;
}

<picture
  className="c0"
  onLoad={[Function]}
>
  <source
    media="(min-width: 600px)"
    srcSet="img-big.jpg"
  />
  <img
    alt="alt"
    className="c1"
    src="img.jpg"
  />
</picture>
`;
